File: LeetCode/0477. Total Hamming Distance/solution.py
"""
46 / 46 test cases passed.
Runtime: 360 ms
Memory Usage: 15.7 MB
"""
class Solution:
    def totalHammingDistance(self, nums: List[int]) -> int:
        n = len(nums)
        ans = 0
        for i in range(30):
            c = 0
            for num in nums:
                c += (num >> i) & 1
            ans += c * (n - c)
        return ans

File: LeetCode/0477. Total Hamming Distance/solution.cpp
/**
 * 46 / 46 test cases passed.
 * Runtime: 36 ms
 * Memory Usage: 18.6 MB 
 */
class Solution {
public:
    int totalHammingDistance(vector<int>& nums) {
        int ans = 0;
        int n = nums.size();
        // the length of the binary number of 10 ^ 9 is 30
        for (int i = 0; i < 30; ++ i) {
            int c = 0;
            for (auto num: nums) {
                c += (num >> i) & 1;
            }
            ans += c * (n - c);
        }
        return ans;
    }
};
